analyzers/tests/SonarAnalyzer.UnitTest/TestCases/FindInsteadOfFirstOrDefault.cs

```diff
@@ -0,0 +1,167 @@
+    {
+        public MyList Fluent() => this;
+    }
+
+    public class HiddenList : List<int>
+    {
+        public int? FirstOrDefault(Func<int, bool> predicate) => null;
+    }
+
+    public static class HelperClass
+    {
+        public static List<int> DoWorkReturnGroup() => null;
+        public static void DoWorkMethodGroup<T>(Func<Func<T, bool>, T> firstOrDefault) { }
+
+        public static bool FilterMethod(int nb) => true;
+    }
+
+    public bool FilterMethod(int nb) => true;
+
+    public void ListBasic(List<int> data)
+    {
+        _ = data.FirstOrDefault(x => true); // Noncompliant {{"Find" method should be used instead of the "FirstOrDefault" extension method.}}
+        //       ^^^^^^^^^^^^^^
+        _ = data.Find(x => true); // Compliant
+
+        _ = data.FirstOrDefault(); // Compliant
+        _ = data.FirstOrDefault(HelperClass.FilterMethod); // Noncompliant
+        //       ^^^^^^^^^^^^^^
+        _ = data.FirstOrDefault(FilterMethod); // Noncompliant
+        //       ^^^^^^^^^^^^^^
+    }
+
+    public void ThroughLinq(List<int> data)
+    {
+        data.Select(x => x * 2).ToList().FirstOrDefault(x => true); // Noncompliant
+        //                               ^^^^^^^^^^^^^^
+        data.Select(x => x * 2).ToList().Find(x => true); // Compliant
+    }
+
+    public void ThroughFunction()
+    {
+        _ = HelperClass.DoWorkReturnGroup().FirstOrDefault(x => true); // Noncompliant
+        //                                  ^^^^^^^^^^^^^^
+        _ = HelperClass.DoWorkReturnGroup().Find(x => true); // Compliant
+    }
+
+    public void ThroughLambda(Func<List<int>> lambda)
+    {
+        _ = lambda().FirstOrDefault(x => true); // Noncompliant
+        //           ^^^^^^^^^^^^^^
+        _ = lambda().Find(x => true); // Compliant
+    }
+
+    public void WithinALambda()
+    {
+        _ = new Func<List<int>, int>(list => list.FirstOrDefault(x => true)); // Noncompliant
+        //                                        ^^^^^^^^^^^^^^
+    }
+
+    public void AsMethodGroup(List<int> data)
+    {
+        HelperClass.DoWorkMethodGroup<int>(data.FirstOrDefault); // FN, this raise as a SimpleAccessMemberExpression
+    }
+
+    public void Miscellaneous(List<int> data)
+    {
+        _ = nameof(Enumerable.FirstOrDefault); // Compliant
+    }
+
+    public static void FirstOrdefaultNotHidden(MyList data)
+    {
+        _ = data.FirstOrDefault(x => true); // Noncompliant
+        //       ^^^^^^^^^^^^^^
+        _ = data.Find(x => true); // Compliant
+    }
+
+    public static void FirstOrdefaultHidden(HiddenList data)
+    {
+        _ = data.FirstOrDefault(x => true); // Compliant
+        _ = data.Find(x => true); // Compliant
+    }
+
+    public static void Nullable(List<int> data = null)
+    {
+        _ = data?.FirstOrDefault(x => true); // Noncompliant
+        //        ^^^^^^^^^^^^^^
+        _ = data?.Find(x => true); // Compliant
+    }
+
+    public static void Ternary(List<int> data)
```

Comment:
Very nitpicky: This contains both ternary and null-coalesce operators. 
You could either rename it to e.g. "SpecialPatterns", OR you could also merge some of the tests.

In general, the testcases look very good, but you do not need this level of granularity IMO, you could spread them in 2-3 methods with a more generic split-basis. (Not talking exclusively about this PR, but in general)

Still, don't waste too much time, either ignore/resolve this or just rename the method :)

Reply:
I just renamed it. 
I like the granularity; it is easier to see which scenario we are testing. But I still need to find the right balance.

Reply:
Sounds good!